toadr3/reports.py

```diff
@@ -0,0 +1,154 @@
+    limit : int | None
+        The maximum number of reports to return.
+
+    Returns
+    -------
+    list[Report]
+        A list of report objects.
+
+    Raises
+    ------
+    ValueError
+        If the query parameters are invalid.
+    toadr3.ToadrException
+        If the request to the VTN fails.
+    """
+    _check_arguments(skip, limit)
+    params = _create_query_parameters(program_id, event_id, client_name, skip, limit)
+
+    headers = {}
+    if access_token is not None:
+        headers["Authorization"] = f"Bearer {access_token.token}"
+
+    while vtn_url.endswith("/"):
+        vtn_url = vtn_url[:-1]
+
+    async with session.get(f"{vtn_url}/reports", params=params, headers=headers) as response:
+        # 400 is start of HTTP error codes
+        if response.status >= 400:  # noqa PLR2004 - Magic value used in comparison
+            match response.status:
+                case 400 | 403 | 500:
+                    json_response = await response.json()  # JSON should be of type Problem schema
+                    message = json_response["title"]
+
+                    if "detail" in json_response:
+                        message = f"{message} - {json_response['detail']}"
+
+                    raise ToadrError(
+                        message,
+                        status_code=response.status,
+                        reason=response.reason,
+                        headers=response.headers,
+                        json_response=json_response,
+                    )
+                case _:
+                    raise RuntimeError(
+                        f"Unexpected response status: {response.status} {response.reason}"
+                    )
+
+        data = await response.json()
+
+        result = []
+        for event in data:
+            result.append(Report(event))
+        return result
+
+
+def _create_query_parameters(
+    program_id: str | None,
+    event_id: str | None,
+    client_name: str | None,
+    skip: int | None,
+    limit: int | None,
+) -> dict[str, str | int]:
+    """Create the query parameters for the get_reports function.
+
+    Parameters
+    ----------
+    program_id : str | None
+        The program ID to filter the events by.
+    event_id : str | None
+        The event ID to filter the events by.
+    client_name : str | None
+        The client name to filter the events by.
+    skip : int | None
+        The number of events to skip (for pagination).
+    limit : int | None
+        The maximum number of events to return.
+    """
+    params = {}
+    if program_id is not None:
+        params["programID"] = program_id
+    if event_id is not None:
+        params["eventID"] = event_id
+    if client_name is not None:
+        params["clientName"] = client_name
+    if skip is not None:
+        params["skip"] = skip
+    if limit is not None:
+        params["limit"] = limit
+    return params
+
+
+def _check_arguments(skip: int | None, limit: int | None):
```

Comment:
It might be nicer to collect all errors, so that the consumer knows about all problems immediately. Now it becomes an iterative process if there are multiple issues.

Reply:
Yes, that sounds very useful. Adding. :)